File: csp/samd21/swift/drivers/Samd21Uart.swift
// HAND-WRITTEN Swift layer-1 driver for the SAMD21 SERCOM-USART -- the Swift projection of
// csp/samd21/csharp/drivers/Samd21Uart.cs, parameterized by a Samd21SercomUsartBinding
// descriptor. Every register offset, field, and width comes from the GENERATED Swift block
// layouts; nothing here is a Swift-local hardware literal. The composed CTRLA words
// (0x40310084 pad2/pad3, 0x40100084 pad0/pad1) are DRIVER-composed from those fields --
// byte-identical to the C# driver's composition -- not emitted.
//
// Requires the Swift `Mmio` namespace (volatile 8/16/32-bit load/store -- the Swift analog
// of C#'s Lamella.Hardware.Mmio, lowered by the AOT tier). Embedded-Swift subset only:
// caseless-enum namespaces + structs, no existentials, no reflection.

/// ONE SAMD21 SERCOM-USART driver for every SERCOM on every SAMD21 board, parameterized by
/// a Samd21SercomUsartBinding. Absolute addresses are computed ONCE in `init` into `let`
/// fields (hot loops touch fields and locals). The SAMD21 needs SUB-WORD MMIO (a Cortex-M0+
/// faults on an unaligned 32-bit access): CLKCTRL is 16-bit at an unaligned offset, STATUS
/// and the PORT mux/config bytes are 8-bit -- reached with write16/read8 per each
/// register's generated *_WIDTH.
///
/// Frame policy (8N1, LSB-first, run-in-standby, 16x oversampling) is DRIVER policy
/// composed from layout fields at init; the composition is pinned by the no-drift gate.
public struct Samd21Uart {
    private let ctrla: UInt32
    private let ctrlb: UInt32
    private let baud: UInt32
    private let intflag: UInt32
    private let syncbusy: UInt32
    private let data: UInt32
    private let binding: Samd21SercomUsartBinding

    /// Binds the driver to one SERCOM-USART wiring; no hardware is touched until `initialize()`.
    public init(_ binding: Samd21SercomUsartBinding) {
        self.binding = binding
        self.ctrla = binding.sercomBase + Samd21SercomUsartLayout.CTRLA_OFF
        self.ctrlb = binding.sercomBase + Samd21SercomUsartLayout.CTRLB_OFF
        self.baud = binding.sercomBase + Samd21SercomUsartLayout.BAUD_OFF
        self.intflag = binding.sercomBase + Samd21SercomUsartLayout.INTFLAG_OFF
        self.syncbusy = binding.sercomBase + Samd21SercomUsartLayout.SYNCBUSY_OFF
        self.data = binding.sercomBase + Samd21SercomUsartLayout.DATA_OFF
    }

    /// Brings the bound SERCOM up as a USART at the binding's rate, 8N1 LSB-first: gates the
    /// APB clock, routes the core clock per the binding's composed GCLK word, muxes TX/RX
    /// (RX input buffer ON), configures, then enables -- each enable-protected write waiting
    /// its SYNCBUSY bit. Idempotent -- safe over a SERCOM the resident firmware already configured.
    public func initialize() {
        // Clock: gate the instance's APB clock, route its core clock, wait GCLK sync.
        let apbcmask = Samd21Instances.PM_BASE + Samd21PmLayout.APBCMASK_OFF
        Mmio.write32(apbcmask, Mmio.read32(apbcmask) | binding.apbcMask)
        Mmio.write16(Samd21Instances.GCLK_BASE + Samd21GclkLayout.CLKCTRL_OFF,
                     UInt16(truncatingIfNeeded: binding.gclkClkctrlValue))
        while (Mmio.read8(Samd21Instances.GCLK_BASE + Samd21GclkLayout.STATUS_OFF)
               & UInt8(truncatingIfNeeded: Samd21GclkLayout.STATUS_SYNCBUSY)) != 0 {}

        // Pinmux: both pins to the binding's function; RX additionally needs INEN (without
        // it the SERCOM never sees the incoming line -- the classic SAMD21 UART gotcha).
        Mmio.write8(binding.pmuxReg, UInt8(truncatingIfNeeded: binding.pmuxPair))
        Mmio.write8(binding.pincfgTxReg, UInt8(truncatingIfNeeded: Samd21PortLayout.PINCFG0_PMUXEN))
        Mmio.write8(binding.pincfgRxReg,
                    UInt8(truncatingIfNeeded: Samd21PortLayout.PINCFG0_PMUXEN
                                              | Samd21PortLayout.PINCFG0_INEN))

        // CTRLA (while disabled): USART with internal clock, the binding's pad routing,
        // LSB-first, run-in-standby; then the plan-derived divisor; then TXEN|RXEN; each
        // write synchronized.
        let ctrlaWord =
            (Samd21SercomUsartLayout.CTRLA_MODE_USART_INTCLK << Samd21SercomUsartLayout.CTRLA_MODE_LSB)
            | Samd21SercomUsartLayout.CTRLA_RUNSTDBY
            | (binding.txpo << Samd21SercomUsartLayout.CTRLA_TXPO_LSB)
            | (binding.rxpo << Samd21SercomUsartLayout.CTRLA_RXPO_LSB)
            | Samd21SercomUsartLayout.CTRLA_DORD
        Mmio.write32(ctrla, ctrlaWord)
        Mmio.write16(baud, UInt16(truncatingIfNeeded: binding.baudDivisor))
        Mmio.write32(ctrlb, Samd21SercomUsartLayout.CTRLB_TXEN | Samd21SercomUsartLayout.CTRLB_RXEN)
        while (Mmio.read32(syncbusy) & Samd21SercomUsartLayout.SYNCBUSY_CTRLB) != 0 {}
        Mmio.write32(ctrla, Mmio.read32(ctrla) | Samd21SercomUsartLayout.CTRLA_ENABLE)
        while (Mmio.read32(syncbusy) & Samd21SercomUsartLayout.SYNCBUSY_ENABLE) != 0 {}
    }

    /// Sends one byte, waiting (bounded) for the Data Register Empty flag first.
    public func writeByte(_ value: UInt8) {
        for _ in 0 ..< 1_000_000 {
            if (Mmio.read8(intflag) & UInt8(truncatingIfNeeded: Samd21SercomUsartLayout.INTFLAG_DRE)) != 0 { break }
        }
        Mmio.write16(data, UInt16(value))
    }

    /// Sends a compile-time string as its low-byte (ASCII) characters -- the greeting path,
    /// allocation-free (StaticString, no managed string).
    public func write(_ text: StaticString) {
        text.withUTF8Buffer { buffer in
            for byte in buffer { writeByte(byte) }
        }
    }

    /// Sends a caller-owned byte buffer -- the neutral `uart.write(buf, len)` shape (the
    /// caller owns the buffer; the driver retains nothing past return; allocation-free).
    public func write(_ bytes: UnsafeBufferPointer<UInt8>) {
        for byte in bytes { writeByte(byte) }
    }

    /// Waits (bounded) until the last frame has left the shift register (INTFLAG.TXC), so a
    /// mode switch or a caller's delay never cuts a frame on the wire.
    public func flush() {
        for _ in 0 ..< 1_000_000 {
            if (Mmio.read8(intflag) & UInt8(truncatingIfNeeded: Samd21SercomUsartLayout.INTFLAG_TXC)) != 0 { return }
        }
    }

    /// 1 when at least one received byte waits (INTFLAG.RXC), else 0.
    public var available: Int {
        (Mmio.read8(intflag) & UInt8(truncatingIfNeeded: Samd21SercomUsartLayout.INTFLAG_RXC)) != 0 ? 1 : 0
    }

    /// Pops one received byte (the DATA field is 9 bits wide), or -1 when the RX register is
    /// empty (the Stream convention: no waiting byte is data, never an error).
    public func readByte() -> Int {
        if (Mmio.read8(intflag) & UInt8(truncatingIfNeeded: Samd21SercomUsartLayout.INTFLAG_RXC)) == 0 { return -1 }
        let raw = UInt32(Mmio.read16(data))
        return Int(raw & Samd21SercomUsartLayout.DATA_DATA)
    }
}

File: csp/samd21/csharp/drivers/Samd21Uart.cs
// The SAMD21 SERCOM-USART driver, in C# over Lamella.Hardware.Mmio -- ONE driver for every
using Lamella.Boards;
using Lamella.Generated;
using Lamella.Hardware;

public sealed class Samd21Uart
{
    private readonly uint _ctrla;
    private readonly uint _ctrlb;
    private readonly uint _baud;
    private readonly uint _intflag;
    private readonly uint _syncbusy;
    private readonly uint _data;
    private readonly Samd21SercomUsartBinding _binding;

    /// <summary>Binds the driver to one SERCOM-USART wiring; no hardware is touched until
    /// <see cref="Init"/>.</summary>
    public Samd21Uart(Samd21SercomUsartBinding binding)
    {
        _binding = binding;
        _ctrla = binding.SercomBase + Samd21SercomUsartLayout.CTRLA_OFF;
        _ctrlb = binding.SercomBase + Samd21SercomUsartLayout.CTRLB_OFF;
        _baud = binding.SercomBase + Samd21SercomUsartLayout.BAUD_OFF;
        _intflag = binding.SercomBase + Samd21SercomUsartLayout.INTFLAG_OFF;
        _syncbusy = binding.SercomBase + Samd21SercomUsartLayout.SYNCBUSY_OFF;
        _data = binding.SercomBase + Samd21SercomUsartLayout.DATA_OFF;
    }

    /// <summary>Brings the bound SERCOM up as a USART at the binding's rate, 8N1 LSB-first:
    /// gates the APB clock, routes the core clock per the binding's composed GCLK word, muxes
    /// the TX/RX pins (RX input buffer ON), configures, then enables -- each enable-protected
    /// write waiting out its SYNCBUSY bit. Idempotent -- safe over a SERCOM the resident
    /// firmware already configured.</summary>
    /// <exception cref="System.InvalidOperationException">The SERCOM's core clock could not be
    /// routed from its generator, or the SERCOM did not finish synchronizing its setup.</exception>
    public void Init()
    {
        // The instance's bus clock, then its core clock. A core clock already running from another
        // generator -- the resident firmware's, say -- is stopped before it moves (DS40001882D
        // 15.6.3.3).
        uint apbcmask = Samd21Instances.PM_BASE + Samd21PmLayout.APBCMASK_OFF;
        Mmio.Write32(apbcmask, Mmio.Read32(apbcmask) | _binding.ApbcMask);
        if (!Samd21GenericClock.Route(_binding.GclkClkctrlValue))
        {
            Invalid("the SERCOM's core clock could not be routed from its generator");
        }

        Mmio.Write8(_binding.PmuxReg, (byte)_binding.PmuxPair);
        Mmio.Write8(_binding.PincfgTxReg, (byte)Samd21PortLayout.PINCFG0_PMUXEN);
        Mmio.Write8(_binding.PincfgRxReg,
            (byte)(Samd21PortLayout.PINCFG0_PMUXEN | Samd21PortLayout.PINCFG0_INEN));

        uint ctrla = (Samd21SercomUsartLayout.CTRLA_MODE_USART_INTCLK
                << (int)Samd21SercomUsartLayout.CTRLA_MODE_LSB)
            | Samd21SercomUsartLayout.CTRLA_RUNSTDBY
            | (_binding.Txpo << (int)Samd21SercomUsartLayout.CTRLA_TXPO_LSB)
            | (_binding.Rxpo << (int)Samd21SercomUsartLayout.CTRLA_RXPO_LSB)
            | Samd21SercomUsartLayout.CTRLA_DORD;
        Mmio.Write32(_ctrla, ctrla);
        Mmio.Write16(_baud, (ushort)_binding.BaudDivisor);
        Mmio.Write32(_ctrlb,
            Samd21SercomUsartLayout.CTRLB_TXEN | Samd21SercomUsartLayout.CTRLB_RXEN);
        WaitSync(Samd21SercomUsartLayout.SYNCBUSY_CTRLB);
        Mmio.Write32(_ctrla, Mmio.Read32(_ctrla) | Samd21SercomUsartLayout.CTRLA_ENABLE);
        WaitSync(Samd21SercomUsartLayout.SYNCBUSY_ENABLE);
    }

    // A bound on each synchronization wait, so a SERCOM that never finishes is reported rather than
    // waited on forever.
    const int WaitBound = 100000;

    void WaitSync(uint bits)
    {
        for (int spin = 0; spin < WaitBound; spin++)
        {
            if ((Mmio.Read32(_syncbusy) & bits) == 0u)
            {
                return;
            }
        }
        Invalid("the SERCOM did not finish synchronizing its setup");
    }

    // The corlib-free tiers bind only System.Exception's constructor, so the precise type is chosen
    // where a corlib is linked, as the family's other drivers do.
    static void Invalid(string why)
    {
#if LAMELLA_CORLIB_LINKED
        throw new System.InvalidOperationException(why);
#else
        throw new System.Exception(why);
#endif
    }

    /// <summary>Sends one byte, waiting (bounded) for the Data Register Empty flag first.</summary>
    public void WriteByte(int value)
    {
        for (int spin = 0; spin < 1000000; spin++)
        {
            if ((Mmio.Read8(_intflag) & (byte)Samd21SercomUsartLayout.INTFLAG_DRE) != 0) break;
        }
        Mmio.Write16(_data, (ushort)(value & 0xFF));
    }

    /// <summary>Sends a string as its low-byte (ASCII) characters.</summary>
    public void Write(string text)
    {
        for (int i = 0; i < text.Length; i++)
        {
            WriteByte(text[i]);
        }
    }

    /// <summary>Waits (bounded) until the last frame has left the shift register
    /// (INTFLAG.TXC), so a mode switch or a caller's delay never cuts a frame on the wire.</summary>
    public void Flush()
    {
        for (int spin = 0; spin < 1000000; spin++)
        {
            if ((Mmio.Read8(_intflag) & (byte)Samd21SercomUsartLayout.INTFLAG_TXC) != 0) return;
        }
    }

    /// <summary>1 when at least one received byte waits (INTFLAG.RXC), else 0.</summary>
    public int Available
    {
        get
        {
            return (Mmio.Read8(_intflag) & (byte)Samd21SercomUsartLayout.INTFLAG_RXC) != 0 ? 1 : 0;
        }
    }

    /// <summary>Pops one received byte (the DATA field is 9 bits wide), or -1 when the RX
    /// register is empty (the Stream convention).</summary>
    public int ReadByte()
    {
        if ((Mmio.Read8(_intflag) & (byte)Samd21SercomUsartLayout.INTFLAG_RXC) == 0) return -1;
        int raw = Mmio.Read16(_data);
        return raw & (int)Samd21SercomUsartLayout.DATA_DATA;
    }
}
